rfcs/text/0014_notification_service.md

```diff
@@ -0,0 +1,401 @@
+- *Source* - a program entity creating a notification event.
+- *Sourcing* - collects notification events created by different internal sources.
+- *Notification* system - abstracts the way notifications are created, stored, configured, and retrieved.  Ensures compatible operation for on-prem and Cloud Kibana versions.
+- *Delivery* - provides API for Kibana UI to retrieve the latest notifications for a user, to update notification status (is_read).
+- *Settings* - provides API to adjust user-specific notification settings (unsubscribe from / subscribe to a source).
+![image](../images/0014/kns_key_elements.png)
+
+### Notification event
+It consists of a message with an associated CTA to show to an end-user.
+*NotificationEvent* interface
+- `recipient_id: string []` - unique identifiers of all the recipients. It might be the same as `elastic_id` (if the env supports it) - the unique user identifier across all the company products. 
+There is currently no way to unambiguously identify the user. WIP [#82725](https://github.com/elastic/kibana/issues/82725)
+- `priority?: 'normal' | 'high' | 'critical'` - to prioritize the delivery in case of async delivery mechanism. An event with high priority is processed first. Shouldn’t be used in UI. Considered as `normal` if not specified. Important messages have `high` priority. `Critical` is used in emergency cases.
+- `created_at: number` - Unix timestamp in UTC timezone when a notification has been created.
+- `expire_at?: number` - Unix timestamp in UTC timezone when notification will be removed from the system. 
+- `is_pinned?: boolean` - a flag to pin a notification at top of the list of notifications. Used by a source to draw a user's attention to a notification.
+- `source_type: string` - source type. the same as source domain name: `cloud`, `alerting`.
+- `group_id?: string` - identifier to associate several notifications in the UI. Notifications are rendered in form of an event timeline in UI.
+- `content: NotificationContent` - see below.
+
+*NotificationContent* interface (see [UI component](https://github.com/elastic/eui/issues/4257))
+- `icon?: { euiIconType: string }` - icon on the left side of the title.
+Compatible with all the types supported by euiIcon type: a predefined EUI icon name, URL to SVG file, data URL for SVG file.
+- `severity?: 'normal' | 'high' | 'critical'` -  type of severity. Used in UI only. Shows as a text after the `source_type` following the format "Alert: Critical".
+- `message: TranslatedContent` - a notification message is subject to i18n. 
+Locale might be changed on the Space level (on the User level in the future).
+Thus, we cannot determine a message locale during notification creation.
+Considering that Kibana notifications might be rendered outside of Kibana UI (in Cloud UI, for example), the message must contain translations for all the supported locales.
+- `title: TranslatedContent & { url?: string }` - notification header.
```

Comment:
Assuming these URLs are for Kibana, the backward compatibility guarantees for URLs don't cross major versions. To side-step this complication, would it make sense for us to drop all Kibana-generated notifications from the prior major version?

Reply:
> To side-step this complication, would it make sense for us to drop all Kibana-generated notifications from the prior major version?

In this case, the user will lose access to *all* notifications created in the previous version.
If we store notifications with the SO mechanism, might it be better to implement migrations for notifications? Plugins can extend notification service migrations with custom logic. We have something similar for `embeddable` plugin https://github.com/restrry/kibana/blob/099e9bc50fafa89301d438ae7112330da5ab421c/src/plugins/kibana_utils/common/persistable_state/index.ts#L60